src/java/htsjdk/samtools/DownsamplingIterator.java

```diff
@@ -123,6 +152,45 @@ private boolean advance() {
         return this.nextRecord != null;
     }
 
+    /**
+     * Buffers reads until either the end of the file is reached or enough reads have been buffered that
+     * downsampling can be performed and the target accuracy achieved.  Once reads have been buffered,
+     * template names are randomly sampled out for discarding until the desired number of reads have
+     * been discarded.
+     *
+     * @return True if one or more reads have been buffered, false otherwise
+     */
+    private boolean bufferNextChunkOfRecords() {
+        final int templatesToRead = (int) Math.ceil(1 / this.targetAccuracy);
+        final Set<String> names = new HashSet<String>();
+        final List<SAMRecord> recs = new ArrayList<SAMRecord>(templatesToRead);
+
+        while (this.underlyingIterator.hasNext() && names.size() < templatesToRead) {
+            final SAMRecord rec = this.underlyingIterator.next();
+            recs.add(rec);
+
+            if (!this.allowSecondaryAlignments && rec.getNotPrimaryAlignmentFlag()) continue;
```

Comment:
I see. the sam-spec calls this flag  "secondary alignment", but the htsjdk
code it's getNotPrimaryAlignmentFlag...perhaps it's time to deprecate and
have a different function?

not in this PR, clearly!



On Fri, Nov 6, 2015 at 5:23 PM, Tim Fennell <notifications@github.com>
wrote:
>
> In src/java/htsjdk/samtools/DownsamplingIterator.java:
>
> > +     * downsampling can be performed and the target accuracy
achieved.  Once reads have been buffered,
> > +     * template names are randomly sampled out for discarding until
the desired number of reads have
> > +     * been discarded.
> > +     *
> > +     * @return True if one or more reads have been buffered, false
otherwise
> > +     */
> > +    private boolean bufferNextChunkOfRecords() {
> > +        final int templatesToRead = (int) Math.ceil(1 /
this.targetAccuracy);
> > +        final Set<String> names = new HashSet<String>();
> > +        final List<SAMRecord> recs = new
ArrayList<SAMRecord>(templatesToRead);
> > +
> > +        while (this.underlyingIterator.hasNext() && names.size() <
templatesToRead) {
> > +            final SAMRecord rec = this.underlyingIterator.next();
> > +            recs.add(rec);
> > +
> > +            if (!this.allowSecondaryAlignments &&
rec.getNotPrimaryAlignmentFlag()) continue;
>
> No - there's understandably some confusion around this.
getNotPrimaryAlignmentFlag() just returns the boolean/flag bit for primary
vs. secondary. There's another method called something like
isSecondaryOrSupplemental() that does what you're thinking.
>
> —
> Reply to this email directly or view it on GitHub.